Mathlib/Topology/Category/Profinite/Nobeling.lean

```diff
@@ -0,0 +1,1754 @@
+  rw [List.chain'_iff_pairwise]
+  have hr : (·>· : α → α → Prop) = (fun a b ↦ a ≥ b ∧ a ≠ b)
+  · ext a b
+    rw [gt_iff_lt, lt_iff_le_and_ne]
+    exact Iff.and Iff.rfl ⟨fun h ↦ h.symm, fun h ↦ h.symm⟩
+  rw [hr, List.pairwise_and_iff]
+  refine ⟨List.chain'_iff_pairwise.mp ?_, Finset.sort_nodup _ _⟩
+  rw [List.chain'_iff_pairwise]
+  exact Finset.sort_sorted _ _
+
+-- TODO: this lemma is purely about lists, and should be moved to the appropriate API file
+theorem GoodProducts.extracted_1
+  (a : I) (as : List I) (ha : List.Chain' (fun x x_1 ↦ x > x_1) (a :: as))
+  (m : List I) (hm : List.Chain' (fun x x_1 ↦ x > x_1) m)
+  (hmas : m ≤ as) : List.Chain' (fun x x_1 ↦ x > x_1) (a :: m) := by
+  cases m with
+  | nil => simp only [List.chain'_singleton]
+  | cons b bs =>
+    apply hm.cons
+    cases as with
+    | nil =>
+      simp only [le_iff_lt_or_eq, or_false] at hmas
+      exact (List.Lex.not_nil_right (·<·) _ hmas).elim
+    | cons a' as =>
+      rw [List.chain'_cons] at ha
+      refine gt_of_gt_of_ge ha.1 ?_
+      rw [le_iff_lt_or_eq] at hmas
+      cases' hmas with hmas hmas
+      · by_contra' hh
+        rw [← not_le] at hmas
+        apply hmas
+        apply le_of_lt
+        exact (List.lt_iff_lex_lt _ _).mp (List.lt.head _ _ hh)
+      · simp only [List.cons.injEq] at hmas
+        rw [ge_iff_le, le_iff_lt_or_eq]
+        exact Or.inr hmas.1
+
+theorem GoodProducts.extracted_2
+  (a : I) (as : List I) (ha : List.Chain' (fun x x_1 ↦ x > x_1) (a :: as)) (c : Products I →₀ ℤ)
+  (hc : (c.support : Set (Products I)) ⊆ {m | m.val ≤ as})
+  (hsm :
+    ∀ (c : ℤ) (x : LocallyConstant ↑(Set.proj C fun x ↦ x ∈ J) ℤ),
+      e (Set.proj C fun x ↦ x ∈ J) a * c • x = c • (e (Set.proj C fun x ↦ x ∈ J) a * x)) :
+  (Finsupp.sum c fun a_1 b ↦ e (Set.proj C fun x ↦ x ∈ J) a * b • Products.eval (Set.proj C fun x ↦ x ∈ J) a_1) ∈
+    Submodule.span ℤ (Products.eval (Set.proj C fun x ↦ x ∈ J) '' {m | m.val ≤ a :: as}) := by
+  apply Submodule.finsupp_sum_mem
+  intro m hm
+  rw [hsm]
+  apply Submodule.smul_mem
+  apply Submodule.subset_span
+  have hmas : m.val ≤ as
+  · apply hc
+    simpa only [Finset.mem_coe, Finsupp.mem_support_iff] using hm
+  refine ⟨⟨a :: m.val, ?_⟩, ⟨?_, ?_⟩⟩
+  · exact GoodProducts.extracted_1 _ _ ha m.val m.prop hmas
+  · simp only [Set.mem_setOf_eq]
+    -- TODO: the remaining proof in this subgoal could be extracted into an API lemma
+    rw [le_iff_lt_or_eq] at hmas ⊢
+    refine hmas.imp ?_ (congr_arg _)
+    intro hmas
+    have haa : ¬(a < a) := gt_irrefl a
+    exact (List.lt_iff_lex_lt _ _).mp
+      (List.lt.tail haa haa ((List.lt_iff_lex_lt _ _).mpr hmas))
+  · simp only [Products.eval, List.map, List.prod_cons]
+
+theorem GoodProducts.spanFin : ⊤ ≤ Submodule.span ℤ (Set.range (eval (C.proj (· ∈ J)))) := by
+  rw [span_iff_products]
+  refine le_trans (spanFinBasis.span C J) ?_
+  rw [Submodule.span_le]
+  intro _ ⟨x, hrfl⟩; cases hrfl -- TODO: why does `intro` not accept a `rfl` pattern here?
```

Comment:
Apparently there is a difference between `intro` and `rintro`. I thought this was gone in Lean 4.
```suggestion
  rintro _ ⟨x, rfl⟩
```